fortuna/hallucination/base.py

```diff
@@ -0,0 +1,335 @@
+        """
+        if self.multicalibrator is None:
+            raise ValueError("`fit` must be called before this method.")
+
+        self.generative_model.to(
+            torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        )
+
+        (
+            scores,
+            embeddings,
+            which_choices,
+        ) = self._compute_scores_embeddings_which_choices(
+            texts=texts, contexts=contexts
+        )
+        if not calibrate:
+            return scores
+
+        embeddings = self.embedding_reduction_fn(embeddings)
+        embeddings = np.concatenate((embeddings, scores[:, None]), axis=1)
+
+        group_scores = self.grouping_model.predict_proba(
+            embeddings=embeddings,
+        )
+        groups = self._get_groups(group_scores)
+
+        return self.multicalibrator.apply_patches(probs=scores, groups=groups)
+
+    def predict(
+        self,
+        texts: Union[List[str], List[List[str]]],
+        contexts: List[str],
+        calibrate: bool = True,
+        probs: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        """
+        A binary prediction for each text given each context.
+        The method returns 1 of a predicted is greater than 0.5, and 0 otherwise.
+
+        Parameters
+        ----------
+        texts: Union[List[str], List[List[str]]]
+            The texts to fit.
+            This may either be a list of strings (e.g. a list of single answers),
+            or a list of lists of strings (e.g. a list of multi-choice answers).
+        contexts: List[str]
+            A list of contexts (e.g. a list of questions).
+        calibrate: bool
+            Whether to calibration the initial probability estimates.
+        probs: Optional[np.ndarray]
+            Predicted probabilities. If these are available, the method does not compute them under the hood.
+
+        Returns
+        -------
+        np.ndarray
+            The predicted probabilities.
+        """
+        if probs is None:
+            probs = self.predict_proba(
+                texts=texts, contexts=contexts, calibrate=calibrate
+            )
+        return (probs >= 0.5).astype(int)
+
+    def _get_groups(self, group_scores: np.ndarray):
+        groups = group_scores >= self._quantiles[None]
+        return np.concatenate((groups, np.ones((groups.shape[0], 1))), axis=1).astype(
+            bool
+        )
+
+    def _compute_scores_embeddings_which_choices(
+        self,
+        texts: Union[List[str], List[List[str]]],
+        contexts: List[str],
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        scores = []
+        embeddings = []
+        which_choices = []
+
+        for text, context in tqdm(zip(texts, contexts)):
+            context_inputs = self.tokenizer(context, return_tensors="pt").to(
+                self.generative_model.device
+            )
+            len_context_inputs = len(context_inputs)
+            if isinstance(text, list):
```

Comment:
correct me if I am wrong, but you have here 2 cases: 1) text is a list of strings 2) text is a string. In the former case, you run over the list and apply some logic L. In the latter case, you also run the same logic but on the text directly. Can you remove the redundant code by something like
```
if isinstance(text, str):
    text = [text]
```
or maybe even better, move the logic to a function?

Reply:
Yes looks possible, I don't know why I didn't.